daffodil-runtime1/src/main/scala/org/apache/daffodil/dsom/CompiledExpression1.scala

```diff
@@ -223,7 +223,7 @@ class DPathCompileInfo(
     with PreSerialization
     with HasSchemaFileLocation {
 
-  def initialize: Unit = {
+  lazy val initialize: Unit = {
     parents
   }
```

Comment:
Why this change, since I don't see any computation being done?

Reply:
I found a StackOverflow [answer](https://stackoverflow.com/questions/7600910/difference-between-function-with-parentheses-and-without) which said that calling nullary methods without an explicit () is no longer optional in Scala 2.13 or 3.0.  BTW, the change wasn't made to annoy people; it was a necessary side effect of another change to improve type inference by allowing eta expansion of Scala functions.  What this tells me is that you shouldn't bother changing `initialize` from a def to a lazy val.  When we upgrade to 2.13, then scalac will enforce consistency between calls and definitions and we can fix any remaining calls without () at that time.

Reply:
This is super helpful. So use () when the method has side effects, but no args. Leave out parens if it is side-effect free. 

Note that a lazy val that only reads things is still considered side-effect free, even though it has this lazy-val once-only behavior. So if you define a def, intending to possibly override with a lazy val, then no parameters is correct, as a lazy val can't have even an empty argument list. 

Also, **_conceptual_** side-effect free applies. Initializing something that happens once as part of setup of some sort is NOT considered a side effect. 

A side-effect is something that has to happen *every* time the method is used. E.g, like assigning something each time, or doing an I/O operation each time.

Reply:
Actually, based on this. We really can't use "-Ywarn-nullary-unit", because it forces us to use () for def that want to be overridden as lazy val.

Reply:
This code is back to being a def with no args.